File: Projects/DMDPDF/PDFBox/pdmodel/graphics/optionalcontent/PDOptionalContentProperties.vb
Imports FinSeA
Imports FinSeA.org.apache.pdfbox.cos
Imports FinSeA.org.apache.pdfbox.pdmodel.common

Namespace org.apache.pdfbox.pdmodel.graphics.optionalcontent

    '/**
    ' * Me class represents the optional content properties dictionary.
    ' *
    ' * @since PDF 1.5
    ' * @version $Revision$
    ' */
    Public Class PDOptionalContentProperties
        Implements COSObjectable

        ''' <summary>
        ''' Enumeration for the BaseState dictionary entry on the "D" dictionary.
        ''' </summary>
        ''' <remarks></remarks>
        Public Structure BaseState

            ''' <summary>
            ''' The "ON" value
            ''' </summary>
            ''' <remarks></remarks>
            Public Shared ReadOnly [ON] = COSName.ON

            ''' <summary>
            ''' The "OFF" value. 
            ''' </summary>
            ''' <remarks></remarks>
            Public Shared ReadOnly [OFF] = COSName.OFF

            '''The "Unchanged" value. */
            Public Shared ReadOnly UNCHANGED = COSName.UNCHANGED

            Private name As COSName

            Friend Sub New(ByVal value As COSName)
                Me.name = value
            End Sub

            ''/**
            '' * Returns the PDF name for the state.
            '' * @return the name of the state
            '' */
            Public Function getName() As COSName
                Return Me.name
            End Function

            ''/**
            '' * Returns the base state represented by the given {@link COSName}.
            '' * @param state the state name
            '' * @return the state enum value
            '' */
            Public Shared Function valueOf(ByVal state As COSName) As BaseState
                If (state Is Nothing) Then
                    Return BaseState.ON
                End If
                Return New BaseState(state) '.getName().ToUpper())
            End Function

            Public Shared Widening Operator CType(ByVal state As COSName) As BaseState
                If (state Is Nothing) Then
                    Return BaseState.ON
                End If
                Return New BaseState(state)  'Return BaseState.valueOf(state.getName().ToUpper())
            End Operator

        End Structure


        Private dict As COSDictionary

        '/**
        ' * Creates a new optional content properties dictionary.
        ' */
        Public Sub New()
            Me.dict = New COSDictionary()
            Me.dict.setItem(COSName.OCGS, New COSArray())
            Me.dict.setItem(COSName.D, New COSDictionary())
        End Sub

        '/**
        ' * Creates a new instance based on a given {@link COSDictionary}.
        ' * @param props the dictionary
        ' */
        Public Sub New(ByVal props As COSDictionary)
            Me.dict = props
        End Sub

        Public Function getCOSObject() As COSBase Implements COSObjectable.getCOSObject
            Return Me.dict
        End Function

        Private Function getOCGs() As COSArray
            Dim ocgs As COSArray = Me.dict.getItem(COSName.OCGS)
            If (ocgs Is Nothing) Then
                ocgs = New COSArray()
                Me.dict.setItem(COSName.OCGS, ocgs) 'OCGs is required
            End If
            Return ocgs
        End Function

        Private Function getD() As COSDictionary
            Dim d As COSDictionary = Me.dict.getDictionaryObject(COSName.D)
            If (d Is Nothing) Then
                d = New COSDictionary()
                Me.dict.setItem(COSName.D, d) 'D is required
            End If
            Return d
        End Function

        '/**
        ' * Returns the optional content group of the given name.
        ' * @param name the group name
        ' * @return the optional content group or null, if there is no such group
        ' */
        Public Function getGroup(ByVal name As String) As PDOptionalContentGroup
            Dim ocgs As COSArray = getOCGs()
            For Each o As COSBase In ocgs
                Dim ocg As COSDictionary = toDictionary(o)
                Dim groupName As String = ocg.getString(COSName.NAME)
                If (groupName.Equals(name)) Then
                    Return New PDOptionalContentGroup(ocg)
                End If
            Next
            Return Nothing
        End Function

        '/**
        ' * Adds an optional content group (OCG).
        ' * @param ocg the optional content group
        ' */
        Public Sub addGroup(ByVal ocg As PDOptionalContentGroup)
            Dim ocgs As COSArray = getOCGs()
            ocgs.add(ocg.getCOSObject())

            'By default, add new group to the "Order" entry so it appears in the user interface
            Dim order As COSArray = getD().getDictionaryObject(COSName.ORDER)
            If (order Is Nothing) Then
                order = New COSArray()
                getD().setItem(COSName.ORDER, order)
            End If
            order.add(ocg)
        End Sub

        '/**
        ' * Returns the collection of all optional content groups.
        ' * @return the optional content groups
        ' */
        Public Function getOptionalContentGroups() As Global.System.Collections.Generic.IEnumerable(Of PDOptionalContentGroup) ' Collection<PDOptionalContentGroup>
            Dim coll As New ArrayList(Of PDOptionalContentGroup)
            Dim ocgs As COSArray = getOCGs()
            For Each base As COSBase In ocgs
                Dim obj As COSObject = base 'Children must be indirect references
                coll.add(New PDOptionalContentGroup(obj.getObject()))
            Next
            Return coll
        End Function

        '/**
        ' * Returns the base state for optional content groups.
        ' * @return the base state
        ' */
        Public Function getBaseState() As BaseState
            Dim d As COSDictionary = getD()
            Dim name As COSName = d.getItem(COSName.BASE_STATE)
            Return BaseState.valueOf(name)
        End Function

        '/**
        ' * Sets the base state for optional content groups.
        ' * @param state the base state
        ' */
        Public Sub setBaseState(ByVal state As BaseState)
            Dim d As COSDictionary = getD()
            d.setItem(COSName.BASE_STATE, state.getName())
        End Sub

        '/**
        ' * Lists all optional content group names.
        ' * @return an array of all names
        ' */
        Public Function getGroupNames() As String()
            Dim ocgs As COSArray = dict.getDictionaryObject(COSName.OCGS)
            Dim size As Integer = ocgs.size()
            Dim groups As String() = Array.CreateInstance(GetType(String), size)
            For i As Integer = 0 To size - 1
                Dim obj As COSBase = ocgs.get(i)
                Dim ocg As COSDictionary = toDictionary(obj)
                groups(i) = ocg.getString(COSName.NAME)
            Next
            Return groups
        End Function

        '/**
        ' * Indicates whether a particular optional content group is found in the PDF file.
        ' * @param groupName the group name
        ' * @return true if the group exists, false otherwise
        ' */
        Public Function hasGroup(ByVal groupName As String) As Boolean
            Dim layers() As String = getGroupNames()
            For Each layer As String In layers
                If (layer.Equals(groupName)) Then
                    Return True
                End If
            Next
            Return False
        End Function

        '/**
        ' * Indicates whether an optional content group is enabled.
        ' * @param groupName the group name
        ' * @return true if the group is enabled
        ' */
        Public Function isGroupEnabled(ByVal groupName As String) As Boolean
            'TODO handle Optional Content Configuration Dictionaries,
            'i.e. OCProperties/Configs

            Dim d As COSDictionary = getD()
            Dim [on] As COSArray = d.getDictionaryObject(COSName.ON)
            If ([on] IsNot Nothing) Then
                For Each o As COSBase In [on]
                    Dim group As COSDictionary = toDictionary(o)
                    Dim name As String = group.getString(COSName.NAME)
                    If (name.Equals(groupName)) Then
                        Return True
                    End If
                Next
            End If

            Dim off As COSArray = d.getDictionaryObject(COSName.OFF)
            If (off IsNot Nothing) Then
                For Each o As COSBase In off
                    Dim group As COSDictionary = toDictionary(o)
                    Dim name As String = group.getString(COSName.NAME)
                    If (name.Equals(groupName)) Then
                        Return False
                    End If
                Next
            End If

            Dim baseState As BaseState = getBaseState()
            Dim enabled As Boolean = Not baseState.Equals(baseState.OFF)
            'TODO What to do with BaseState.Unchanged?
            Return enabled
        End Function

        Private Function toDictionary(ByVal o As COSBase) As COSDictionary
            If (TypeOf (o) Is COSObject) Then
                Return DirectCast(o, COSObject).getObject()
            Else
                Return o
            End If
        End Function

        '/**
        ' * Enables or disables an optional content group.
        ' * @param groupName the group name
        ' * @param enable true to enable, false to disable
        ' * @return true if the group already had an on or off setting, false otherwise
        ' */
        Public Function setGroupEnabled(ByVal groupName As String, ByVal enable As Boolean) As Boolean
            Dim d As COSDictionary = getD()
            Dim [on] As COSArray = d.getDictionaryObject(COSName.ON)
            If ([on] Is Nothing) Then
                [on] = New COSArray()
                d.setItem(COSName.ON, [on])
            End If
            Dim off As COSArray = d.getDictionaryObject(COSName.OFF)
            If (off Is Nothing) Then
                off = New COSArray()
                d.setItem(COSName.OFF, off)
            End If

            Dim found As Boolean = False
            If (enable) Then
                For Each o As COSBase In off
                    Dim group As COSDictionary = toDictionary(o)
                    Dim name As String = group.getString(COSName.NAME)
                    If (name.Equals(groupName)) Then
                        'enable group
                        off.remove(group)
                        [on].add(group)
                        found = True
                        Exit For
                    End If
                Next
            Else
                For Each o As COSBase In [on]
                    Dim group As COSDictionary = toDictionary(o)
                    Dim name As String = group.getString(COSName.NAME)
                    If (name.Equals(groupName)) Then
                        'disable group
                        [on].remove(group)
                        off.add(group)
                        found = True
                        Exit For
                    End If
                Next
            End If
            If (Not found) Then
                Dim ocg As PDOptionalContentGroup = getGroup(groupName)
                If (enable) Then
                    [on].add(ocg.getCOSObject())
                Else
                    off.add(ocg.getCOSObject())
                End If
            End If
            Return found
        End Function

    End Class

End Namespace

File: Projects/FinSeAPDF/PDFBox/pdmodel/graphics/optionalcontent/PDOptionalContentProperties.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.pdfbox.pdmodel.graphics.optionalcontent;

import java.util.Collection;

import org.apache.pdfbox.cos.COSArray;
import org.apache.pdfbox.cos.COSBase;
import org.apache.pdfbox.cos.COSDictionary;
import org.apache.pdfbox.cos.COSName;
import org.apache.pdfbox.cos.COSObject;
import org.apache.pdfbox.pdmodel.common.COSObjectable;

/**
 * This class represents the optional content properties dictionary.
 *
 * @since PDF 1.5
 * @version $Revision$
 */
public class PDOptionalContentProperties implements COSObjectable
{

    /**
     * Enumeration for the BaseState dictionary entry on the "D" dictionary.
     */
    public static enum BaseState
    {

        /** The "ON" value. */
        ON(COSName.ON),
        /** The "OFF" value. */
        OFF(COSName.OFF),
        /** The "Unchanged" value. */
        UNCHANGED(COSName.UNCHANGED);

        private COSName name;

        private BaseState(COSName value)
        {
            Me.name = value;
        }

        /**
         * Returns the PDF name for the state.
         * @return the name of the state
         */
        public COSName getName()
        {
            return Me.name;
        }

        /**
         * Returns the base state represented by the given {@link COSName}.
         * @param state the state name
         * @return the state enum value
         */
        public static BaseState valueOf(COSName state)
        {
            if (state Is Nothing)
            {
                return BaseState.ON;
            }
            return BaseState.valueOf(state.getName().toUpperCase());
        }

    }

    private COSDictionary dict;

    /**
     * Creates a new optional content properties dictionary.
     */
    public PDOptionalContentProperties()
    {
        Me.dict = new COSDictionary();
        Me.dict.setItem(COSName.OCGS, new COSArray());
        Me.dict.setItem(COSName.D, new COSDictionary());
    }

    /**
     * Creates a new instance based on a given {@link COSDictionary}.
     * @param props the dictionary
     */
    public PDOptionalContentProperties(COSDictionary props)
    {
        Me.dict = props;
    }

    /** {@inheritDoc} */
    public COSBase getCOSObject()
    {
        return Me.dict;
    }

    private COSArray getOCGs()
    {
        COSArray ocgs = (COSArray)Me.dict.getItem(COSName.OCGS);
        if (ocgs Is Nothing)
        {
            ocgs = new COSArray();
            Me.dict.setItem(COSName.OCGS, ocgs); //OCGs is required
        }
        return ocgs;
    }

    private COSDictionary getD()
    {
        COSDictionary d = (COSDictionary)Me.dict.getDictionaryObject(COSName.D);
        if (d Is Nothing)
        {
            d = new COSDictionary();
            Me.dict.setItem(COSName.D, d); //D is required
        }
        return d;
    }

    /**
     * Returns the optional content group of the given name.
     * @param name the group name
     * @return the optional content group or null, if there is no such group
     */
    public PDOptionalContentGroup getGroup(String name)
    {
        COSArray ocgs = getOCGs();
        for (COSBase o : ocgs)
        {
            COSDictionary ocg = toDictionary(o);
            String groupName = ocg.getString(COSName.NAME);
            if (groupName.equals(name))
            {
                return new PDOptionalContentGroup(ocg);
            }
        }
        return null;
    }

    /**
     * Adds an optional content group (OCG).
     * @param ocg the optional content group
     */
    public void addGroup(PDOptionalContentGroup ocg)
    {
        COSArray ocgs = getOCGs();
        ocgs.add(ocg.getCOSObject());

        //By default, add new group to the "Order" entry so it appears in the user interface
        COSArray order = (COSArray)getD().getDictionaryObject(COSName.ORDER);
        if (order Is Nothing)
        {
            order = new COSArray();
            getD().setItem(COSName.ORDER, order);
        }
        order.add(ocg);
    }

    /**
     * Returns the collection of all optional content groups.
     * @return the optional content groups
     */
    public Collection<PDOptionalContentGroup> getOptionalContentGroups()
    {
        Collection<PDOptionalContentGroup> coll = new java.util.ArrayList<PDOptionalContentGroup>();
        COSArray ocgs = getOCGs();
        for (COSBase base : ocgs)
        {
            COSObject obj = (COSObject)base; //Children must be indirect references
            coll.add(new PDOptionalContentGroup((COSDictionary)obj.getObject()));
        }
        return coll;
    }

    /**
     * Returns the base state for optional content groups.
     * @return the base state
     */
    public BaseState getBaseState()
    {
        COSDictionary d = getD();
        COSName name = (COSName)d.getItem(COSName.BASE_STATE);
        return BaseState.valueOf(name);
    }

    /**
     * Sets the base state for optional content groups.
     * @param state the base state
     */
    public void setBaseState(BaseState state)
    {
        COSDictionary d = getD();
        d.setItem(COSName.BASE_STATE, state.getName());
    }

    /**
     * Lists all optional content group names.
     * @return an array of all names
     */
    public String[] getGroupNames()
    {
        COSArray ocgs = (COSArray)dict.getDictionaryObject(COSName.OCGS);
        int size = ocgs.size();
        String[] groups = new String[size];
        for (int i = 0; i < size; i++)
        {
            COSBase obj = (COSBase)ocgs.get(i);
            COSDictionary ocg = toDictionary(obj);
            groups(i) = ocg.getString(COSName.NAME);
        }
        return groups;
    }

    /**
     * Indicates whether a particular optional content group is found in the PDF file.
     * @param groupName the group name
     * @return true if the group exists, false otherwise
     */
    public boolean hasGroup(String groupName)
    {
        String[] layers = getGroupNames();
        for (String layer : layers)
        {
            if (layer.equals(groupName))
            {
                return true;
            }
        }
        return false;
    }

    /**
     * Indicates whether an optional content group is enabled.
     * @param groupName the group name
     * @return true if the group is enabled
     */
    public boolean isGroupEnabled(String groupName)
    {
        //TODO handle Optional Content Configuration Dictionaries,
        //i.e. OCProperties/Configs

        COSDictionary d = getD();
        COSArray on = (COSArray)d.getDictionaryObject(COSName.ON);
        if (on IsNot Nothing)
        {
            for (COSBase o : on)
            {
                COSDictionary group = toDictionary(o);
                String name = group.getString(COSName.NAME);
                if (name.equals(groupName))
                {
                    return true;
                }
            }
        }

        COSArray off = (COSArray)d.getDictionaryObject(COSName.OFF);
        if (off IsNot Nothing)
        {
            for (COSBase o : off)
            {
                COSDictionary group = toDictionary(o);
                String name = group.getString(COSName.NAME);
                if (name.equals(groupName))
                {
                    return false;
                }
            }
        }

        BaseState baseState = getBaseState();
        boolean enabled = !baseState.equals(BaseState.OFF);
        //TODO What to do with BaseState.Unchanged?
        return enabled;
    }

    private COSDictionary toDictionary(COSBase o)
    {
        if (o instanceof COSObject)
        {
            return (COSDictionary)((COSObject)o).getObject();
        }
        else
        {
            return (COSDictionary)o;
        }
    }

    /**
     * Enables or disables an optional content group.
     * @param groupName the group name
     * @param enable true to enable, false to disable
     * @return true if the group already had an on or off setting, false otherwise
     */
    public boolean setGroupEnabled(String groupName, boolean enable)
    {
        COSDictionary d = getD();
        COSArray on = (COSArray)d.getDictionaryObject(COSName.ON);
        if (on Is Nothing)
        {
            on = new COSArray();
            d.setItem(COSName.ON, on);
        }
        COSArray off = (COSArray)d.getDictionaryObject(COSName.OFF);
        if (off Is Nothing)
        {
            off = new COSArray();
            d.setItem(COSName.OFF, off);
        }

        boolean found = false;
        if (enable)
        {
            for (COSBase o : off)
            {
                COSDictionary group = toDictionary(o);
                String name = group.getString(COSName.NAME);
                if (name.equals(groupName))
                {
                    //enable group
                    off.remove(group);
                    on.add(group);
                    found = true;
                    break;
                }
            }
        }
        else
        {
            for (COSBase o : on)
            {
                COSDictionary group = toDictionary(o);
                String name = group.getString(COSName.NAME);
                if (name.equals(groupName))
                {
                    //disable group
                    on.remove(group);
                    off.add(group);
                    found = true;
                    break;
                }
            }
        }
        if (!found)
        {
            PDOptionalContentGroup ocg = getGroup(groupName);
            if (enable)
            {
                on.add(ocg.getCOSObject());
            }
            else
            {
                off.add(ocg.getCOSObject());
            }
        }
        return found;
    }


}